src/runtime-semantics/RegExp.mjs

```diff
@@ -0,0 +1,777 @@
+        }
+        throw new Error('unreachable');
+      });
+
+      return combinedCharSet(charSets);
+    }
+
+    // 21.2.2.15.1 #sec-runtime-semantics-characterrange-abstract-operation
+    function CharacterRange(A, B) {
+      // 1. Assert: A and B each contain exactly one character.
+      Assert(typeof A === 'number' && typeof B === 'number');
+      const i = A;
+      const j = B;
+      Assert(i <= j);
+      const set = new Set();
+      for (let codePoint = A; codePoint <= B; codePoint += 1) {
+        set.add(Canonicalize(String.fromCodePoint(codePoint)));
+      }
+      return function testCharacterRange(cc) {
+        return set.has(cc);
+      };
+    }
+
+    // 21.2.2.19 #sec-classescape
+    function Evaluate_ClassEscape(ClassEscape) {
+      if (ClassEscape.subtype === 'b') {
+        return '\b'.charCodeAt(0);
+      }
+
+      if (ClassEscape.subtype === '-') {
+        return '-'.charCodeAt(0);
+      }
+
+      if (ClassEscape.subtype === 'CharacterEscape') {
+        return Evaluate_CharacterEscape(ClassEscape.CharacterEscape);
+      }
+
+      if (ClassEscape.subtype === 'CharacterClassEscape') {
+        return Evaluate_CharacterClassEscape(ClassEscape.CharacterClassEscape);
+      }
+
+      throw new Error('unreachable');
+    }
+
+    function singleCharSet(char) {
+      char = Canonicalize(String.fromCodePoint(char));
+      return function testSingleCharSet(cc) {
+        return char === cc;
+      };
+    }
+
+    function allCharSet() {
+      return function testAllCharSet() {
+        return true;
+      };
+    }
+
+    function noLineTerminatorCharSet() {
+      return function testNoLineTerminatorCharSet(cc) {
+        return !isLineTerminator(cc);
+      };
+    }
+
+    function numberCharSet() {
+      return function testNumberCharSet(cc) {
+        return /[0-9]/.test(cc);
+      };
+    }
+
+    function whitespaceCharSet() {
+      return function testWhitespaceCharSet(cc) {
+        return isStrWhiteSpaceChar(cc);
+      };
+    }
+
+    function emptyCharSet() {
+      return function testEmptyCharSet() {
+        return false;
+      };
+    }
+
+    function invertCharSet(charSet) {
+      return function testInvertCharSet(cc) {
+        return !charSet(cc);
+      };
+    }
+
+    function combinedCharSet(charSets) {
+      return function testCombinedCharSet(cc) {
+        return charSets.some((charSet) => charSet(cc));
+      };
+    }
+
+    function classRangeAtomCharSet(classRange) {
+      const classAtom = getClassAtom(classRange);
+      if (typeof classAtom === 'function') {
+        return classAtom;
+      }
+      return singleCharSet(classAtom);
+    }
+
+    function getClassAtom(ClassAtom) {
+      if (ClassAtom.subtype === 'character') {
+        return ClassAtom.character;
+      } else {
+        return Evaluate_ClassEscape(ClassAtom.ClassEscape);
+      }
+    }
+  };
+}
+
+class MatchResult {}
+
+export class State extends MatchResult {
```

Comment:
this should go at the top and have a `#sec-xyz` link to where it comes from in the spec